org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/StateBasedContainerManager.java

```diff
@@ -84,12 +84,18 @@ protected IAllContainersState getState(IResourceDescriptions resourceDescription
 	protected IContainer createContainer(String handle, IResourceDescriptions resourceDescriptions) {
 		IAllContainersState state = getState(resourceDescriptions);
 		IContainerState containerState = new ContainerState(handle, state);
-		StateBasedContainer result = new StateBasedContainer(resourceDescriptions, containerState);
+		
+		StateBasedContainer result = createStateBasedContainer(resourceDescriptions, containerState);
```

Comment:
What if we move the `if (.. instanceof ..)` below also into the createStateBasedContainer method, rename that one to an overload of `createContainer` and let it return `IContainer` instead of an implementation type? That should offer more flexibility for clients.